rest-api/tools/ehr_upload.py

```diff
@@ -1,97 +1,134 @@
+      'gid)s' % {
+        'id': SPREADSHEET_ID,
+        'gid': "0",
+        }
 
 
-def get_participants_without_site_pairing(session, organization, source_bucket, destination_bucket):
-  sql = """
-    select p.participant_id from participant p
+def get_sql(organization):
+  site_pairing_sql = """
+    select p.participant_id, google_group from participant p left join site s on
+    p.site_id = s.site_id
     left join participant_summary summary on p.participant_id = summary.participant_id
-    where p.site_id is NULL and p.organization_id in (
-    select organization_id from organization where external_id = '{}')
+    where p.organization_id in (select organization_id from organization where external_id = '{}')
+    and s.google_group is not null
     and summary.consent_for_electronic_health_records = 1
     and summary.consent_for_study_enrollment = 1;
-   """.format(organization)
-  cursor = session.execute(text(sql))
-  try:
+     """.format(organization)
+
+  no_site_pairing_sql = """
+      select p.participant_id from participant p
+      left join participant_summary summary on p.participant_id = summary.participant_id
+      where p.site_id is NULL and p.organization_id in (
+      select organization_id from organization where external_id = '{}')
+      and summary.consent_for_electronic_health_records = 1
+      and summary.consent_for_study_enrollment = 1;
+     """.format(organization)
+  return site_pairing_sql, no_site_pairing_sql
+
+
+def _fetch_csv_data():
+  response = urllib2.urlopen(FILE_URL)
+  return csv.DictReader(StringIO.StringIO(response.read()))
+
+
+def _ensure_buckets(hpo_data):
+  """Some organizations aggregating org ID is responsible for consent verification. This is a
+  parent/child relationship(not seen elsewhere in the RDR). If bucket name is blank it is safe to
+  assume the parent org (aggregate org id) bucket is to be used."""
+  for _, _dict in hpo_data.items():
+    if _dict['bucket'] == '':
+      parent = _dict['aggregate_id']
+      _dict['bucket'] = hpo_data[parent]['bucket']
+
+
+def read_hpo_report(csv_reader):
+  hpo_data = {}
+
+  for row in csv_reader:
+    bucket_name = row.get(BUCKET_NAME)
+    aggregate_id = row.get(AGGREGATING_ORG_ID)
+    org_id = row.get(ORG_ID)
+    org_status = row.get(ORG_STATUS)
+
+    if org_status == 'Active':
+      hpo_data[org_id] = {'aggregate_id': aggregate_id, 'bucket': bucket_name}
+
+  _ensure_buckets(hpo_data)
+  return hpo_data
+
+
+def sync_ehr_consents():
+  csv_reader = _fetch_csv_data()
+  csv_reader.next()
+  hpo_data = read_hpo_report(csv_reader)
+  logging.info('Reading data complete, beginning sync...')
+  for org, data in hpo_data.items()[:2]:
+    site_paired_sql, no_paired_sql = get_sql(org)
+    logging.info('syncing participants for {}'.format(org))
+    run_sql(data['bucket'], site_paired_sql, no_paired_sql)
+
+
+def run_gsutil(gsutil):
+  gsutil = shlex.split(str(gsutil))
```

Comment:
I think I've made this comment elsewhere, but surely it would be less error prone to pass an array of tokens here rather than encoding the whole thing as a string and splitting it later.

Reply:
Yes. Create an array of the commandline tokens directly and skip this intermediate step (split just returns an array of tokens), e.g.


```
["gsutil", "cp", "gs://my_path", "dest"]
```

Reply:
shlex.split is the commonly accepted Pythonic way of passing commands to Popen().

Reply:
I'm willing to be wrong, but I'm going to need a really good citation to believe that assertion.

In cases where you are starting with an arbitrary input string - sure; that's not what's going on here though. You've built all of the input parameters in code.

(this isn't blocking the review, mostly a style comment)

Reply:
https://docs.python.org/2.7/library/shlex.html#shlex.split
https://www.kevinlondon.com/2015/07/26/dangerous-python-functions.html

Reply:
I don't see any direct relevance to this situation in either of those articles. The Popen docs mention shlex but mainly as a note that it can help to properly tokenize inputs - the examples at the bottom all use direct arrays: https://docs.python.org/2/library/subprocess.html#replacing-older-functions-with-the-subprocess-module

Anyways, fine to leave as-is, just seems less error prone to me to write (simplified, but equivalent by my understanding):

```
x = ["a", "b", "c"]
```

rather than

```
x_str = "a b c"
x = x_str.split(" ")
```